File: bench/compare/run_gradmem.jl
# Peak GPU memory + time for the DERIVATIVE OF THE FORWARD PASS (white noise xi -> correlated
# field) via GraphGP.jl's ANALYTIC adjoint generate_grad_xi — no materialised (M,k+1,k+1) tensor,
# no autodiff tape. Compare to run_gradmem.py (JAX autodiff). Reports CUDA.jl per-call device
# allocation and the device high-water (peak) usage.
#
#   julia --project=bench/compare run_gradmem.jl N [K]
using GraphGP, CUDA, KernelAbstractions, Random, Printf

N = length(ARGS) >= 1 ? parse(Int, ARGS[1]) : 200_000
K = length(ARGS) >= 2 ? parse(Int, ARGS[2]) : 30
D, n0 = 3, 256

rng = MersenneTwister(0)
pts = randn(rng, N, D)
bins0, vals0 = rbf_kernel(1.0, 0.3, 1e-4, 1e1, 200; jitter = 1e-3)
bins = Float32.(bins0); vals = Float32.(vals0)            # f32 production path
prob = build_graph_ka(CuArray(pts), n0, K, CuArray(bins), CuArray(vals); backend = CUDABackend())
w = CuArray(randn(Float32, N))

dev_used() = (CUDA.total_memory() - CUDA.free_memory()) / 1e6   # MB resident on device

xi = CuArray(randn(Float32, N))

# Forward generate (baseline vs the CUDA extension's forward generate).
CUDA.reclaim(); generate(prob, xi); CUDA.reclaim()
falloc = CUDA.@allocated generate(prob, xi)
ft = CUDA.@elapsed generate(prob, xi)

# Derivatives of the forward pass (analytic adjoints; the CUDA extension can't do these at all).
CUDA.reclaim(); generate_grad_xi(prob, w); CUDA.reclaim()
galloc = CUDA.@allocated generate_grad_xi(prob, w)
gt = CUDA.@elapsed generate_grad_xi(prob, w)

CUDA.reclaim(); generate_grad_vals(prob, xi, w); CUDA.reclaim()
valloc = CUDA.@allocated generate_grad_vals(prob, xi, w)
vt = CUDA.@elapsed generate_grad_vals(prob, xi, w)

@printf("GRADMEM julia-gpu N=%d K=%d | forward: alloc=%.1f MB %.1f ms | d/dxi: alloc=%.1f MB %.1f ms | d/dcov_vals: alloc=%.1f MB %.1f ms\n",
    N, K, falloc / 1e6, 1e3ft, galloc / 1e6, 1e3gt, valloc / 1e6, 1e3vt)

File: bench/compare/run_gradmem.py
"""Peak GPU memory of the graphgp CUDA EXTENSION (the fair baseline) on the FORWARD pass
(white noise xi -> correlated field). The extension is forward-only — it has no autodiff — so the
DERIVATIVES of the forward pass (d/dxi, d/dcov_vals) cannot be computed by it at all; those are a
GraphGP.jl-unique capability (run_gradmem.jl). We therefore baseline forward memory here and mark
derivatives as unsupported.

    python run_gradmem.py N [K]

On-demand allocator so peak is real (not XLA's 75% preallocation).
"""
import json
import os
import sys

os.environ["XLA_PYTHON_CLIENT_PREALLOCATE"] = "false"
os.environ.setdefault("JAX_PLATFORMS", "cuda")

import time

import jax
import jax.numpy as jnp
import jax.random as jr
import graphgp as gp

N = int(sys.argv[1]) if len(sys.argv) > 1 else 200_000
K = int(sys.argv[2]) if len(sys.argv) > 2 else 30
D, n0 = 3, 256
dev = jax.devices()[0]

key = jr.key(0)
kp, kx = jr.split(key, 2)
points = jr.normal(kp, (N, D), dtype=jnp.float32)
graph = gp.build_graph(points, n0=n0, k=K, cuda=True)
cov = gp.extras.rbf_kernel(variance=1.0, scale=0.3, r_min=1e-4, r_max=1e1, n_bins=200, jitter=1e-3)
cov = (jnp.asarray(cov[0], jnp.float32), jnp.asarray(cov[1], jnp.float32))
xi = jr.normal(kx, (N,), dtype=jnp.float32)


def peak_mb():
    return (dev.memory_stats() or {}).get("peak_bytes_in_use", 0) / 1e6


# Forward generate via the CUDA extension (cuda=True). No jit needed; the ext is a custom call.
try:
    f = lambda x: gp.generate(graph, cov, x, cuda=True)
    jax.block_until_ready(f(xi))                  # warm
    t0 = time.perf_counter()
    jax.block_until_ready(f(xi))
    dt = time.perf_counter() - t0
    fwd = {"peak_mb": round(peak_mb(), 1), "ms": round(1e3 * dt, 1), "ok": True}
except Exception as e:
    fwd = {"peak_mb": None, "ms": None, "ok": False, "err": type(e).__name__}

res = {"path": "cuda-ext-gpu", "N": N, "K": K,
       "forward_generate": fwd,
       "d_dxi": "unsupported (extension has no autodiff)",
       "d_dvals": "unsupported (extension has no autodiff)"}
print("GRADMEM " + json.dumps(res))
